src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java

```diff
@@ -0,0 +1,303 @@
+    }
+
+    /**
+     * @return Whether this tab's parent GUI is the current GUI
+     */
+    public boolean isActiveTab(Class<? extends GuiContainer> guiContainer)
+    {
+        return guiContainer != null && guiContainer.equals(getTargetGui());
+    }
+
+    /**
+     * @return The tabindex / 12
+     */
+    public int getTabPage()
+    {
+        return getTabIndex() / 12;
+    }
+
+    /**
+     * @return The localized name for this tab
+     */
+    public String getLocalizedName()
+    {
+        return "guiGroup." + name;
+    }
+
+    public ResourceLocation getIconResLoc()
+    {
+        return iconResLoc;
+    }
+
+    public ItemStack getIconStack()
+    {
+        return iconStack;
+    }
+
+    /**
+     * @return A list of the appended tabs for the given GuiContainer
+     */
+    public static List<GuiTab> getTabsForGui(Class<? extends GuiContainer> guiContainer)
+    {
+        return tabRegistry.containsKey(guiContainer) ? tabRegistry.get(guiContainer) : new ArrayList<>();
+    }
+
+    /**
+     * Adds the GuiTab to GuiContainer
+     *
+     * @param parentContainer The GuiContainer to add a tab too
+     * @return The GuiTab
+     */
+    public GuiTab addTo(Class<? extends GuiContainer> parentContainer)
+    {
+        if (parentContainer.equals(GuiContainerCreative.class))
+        {
+            FMLLog.log.warn("Cannot add tabs to creative inventory. Use CreativeTabs instead");
+            return this;
+        }
+        else if (parentContainer.equals(InventoryEffectRenderer.class))
+        {
+            FMLLog.log.warn(name + " tab has a parent that is ambitious with the creative inventory. cannot proceed with registry.");
+            return this;
+        }
+
+        this.parentContainer = parentContainer;
+        tabRegistry.put(parentContainer, this);
+        return this;
+    }
+
+    /**
+     * Adds the GuiTab to GuiContainers
+     *
+     * @param parentContainers The GuiContainers to add a tab too
+     * @return The GuiTab
+     */
+    public GuiTab addTo(Class<? extends GuiContainer>... parentContainers)
+    {
+        Arrays.stream(parentContainers).forEach(this::addTo);
+        return this;
+    }
+
+    /**
+     * Sets the GuiContainer this tab should show on as the main tab.
+     *
+     * @param targetGui The target gui to shown on
+     * @return The GuiTab instance
+     */
+    public GuiTab setTargetGui(Class<? extends GuiContainer> targetGui)
+    {
+        this.targetGui = targetGui;
+        return this;
+    }
+
+    /**
+     * Process interaction when the tab is clicked. should be used to send a packet to the server and open a GUI
+     *
+     * @param guiContainer The parent gui
+     */
+    public void onTabClicked(GuiContainer guiContainer)
+    {
+        int size = getTabsForGui(guiContainer.getClass()).size();
+        if (size > 0)
+        {
+            for (int i = 0; size > i; i++)
```

Comment:
You should use an iterator or a stream for this instead.

Reply:
readable enough ?
```java
       if (size > 0)
        {
            IntStream.range(0, size).filter(i ->
                    !getTabsForGui(targetGui).contains(getTabsForGui(guiContainer.getClass()).get(i))
            ).forEach(i ->
                    getTabsForGui(guiContainer.getClass()).get(i).addTo(targetGui)
            );
        }
```